src/libraries/System.IO.FileSystem/tests/Directory/SymbolicLinks.cs

```diff
@@ -0,0 +1,39 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace System.IO.Tests
+{
+    public class Directory_SymbolicLinks : BaseSymbolicLinks
+    {
+        [Fact]
+        public void EnumerateDirectories_LinksWithCycles_ShouldNotThrow()
+        {
+            DirectoryInfo testDirectory = CreateDirectorySymbolicLinkToItself();
+
+            // Windows differentiates between dir symlinks and file symlinks
+            int expected = PlatformDetection.IsWindows ? 1 : 0;
+            Assert.Equal(expected, Directory.EnumerateDirectories(testDirectory.FullName).Count());
+        }
+
+        [Fact]
+        public void EnumerateFiles_LinksWithCycles_ShouldNotThrow()
+        {
+            DirectoryInfo testDirectory = CreateDirectorySymbolicLinkToItself();
+
+            // Windows differentiates between dir symlinks and file symlinks
+            int expected = PlatformDetection.IsWindows ? 0 : 1;
+            Assert.Equal(expected, Directory.EnumerateFiles(testDirectory.FullName).Count());
+        }
+
+        [Fact]
+        public void EnumerateFileSystemEntries_LinksWithCycles_ShouldNotThrow()
+        {
+            DirectoryInfo testDirectory = CreateDirectorySymbolicLinkToItself();
```

Comment:
all tests being "symlink to itself" is not really exercising any looping issues that emerge from creating symlink to any directory in its parents... like: dir `a` has a symlink `x` pointing to dir `b`, that has subdir `c` which has another symlink `y` pointing to dir `a`

Reply:
Interesting, thanks for the feedback.

I'll add both tests: a symlink to itself, and a symlink to its own parent folder. Maybe this uncovers additional scenarios we weren't aware of.

Reply:
I guess something like this should suffice:
* `link -> link` (what we already have)
* `A -> linkToA`
* `A -> linkToB -> B -> linkToA`

Reply:
> * A -> linkToB -> B -> linkToA

where the only common parent of `A` and `B`  is `/` (e..g `/tmp/A/` and `/var/tmp/B/`)

Reply:
Let's keep this in mind for the future. Right now the only case I want to tackle is a symlink to itself. The case where there is symlink cycle to a parent folder can be rediscussed when we add the follow symlinks feature. We will also have to keep in mind that if we recursively access folders, we end up in an infinite loop.